### PR TITLE
Serial hotplug

### PR DESCRIPTION
**Pull request by xutian**
This PR is a update for fyang's PR#2134

### CODE REVIEW

**Comment by xutian** [comment]
ID : 1157458 1157436

**Comment by FengYang** [comment]
Thanks @xutian

**Comment by vipmike007** [comment]
finally:
         env["serial_file_transfer_start"] = False		         env["serial_file_transfer_start"] = False

The value of serial_file_transfer_start are same before transfer start and transfer complete .during testing ,we might now know whether started or not .Is it better to change the value to another one

**Comment by CongSmile** [comment]
Test failed.

  Traceback (most recent call last):
    File "/home/serial/staf-kvm-devel/autotest-devel/client/shared/test.py", line 823, in _call_test_function
      return func(*args, **dargs)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/shared/test.py", line 291, in execute
      postprocess_profiled_run, args, dargs)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/shared/test.py", line 209, in _call_run_once
      *args, **dargs)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/shared/test.py", line 313, in run_once_profiling
      self.run_once(*args, **dargs)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virt.py", line 145, in run_once
      params = env_process.preprocess(self, params, env)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/shared/error.py", line 141, in new_fn
      return fn(*args, **kwargs)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virttest/env_process.py", line 807, in preprocess
      process(test, params, env, preprocess_image, preprocess_vm)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virttest/env_process.py", line 526, in process
      _call_vm_func()
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virttest/env_process.py", line 495, in _call_vm_func
      vm_func(test, vm_params, env, vm_name)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virttest/env_process.py", line 175, in preprocess_vm
      migration_exec_cmd=params.get("migration_exec_cmd_dst"))
    File "/home/serial/staf-kvm-devel/autotest-devel/client/shared/error.py", line 141, in new_fn
      return fn(*args, **kwargs)
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virttest/qemu_vm.py", line 2505, in create
      self.create_serial_console()
    File "/home/serial/staf-kvm-devel/autotest-devel/client/tests/virt/virttest/qemu_vm.py", line 1969, in create_serial_console
      raise virt_vm.VMConfigMissingError(self.name, "isa_serial")
  VMConfigMissingError: Missing config 'isa_serial' for VM virt-tests-vm1

**Comment by xutian** [comment]
@CongSmile  it looks it's your tree issue, I have test with latest code everything works as expect and not meet above issue.
Please update your tree and apply patch to retry it.

Thanks,
@xutian

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -431,6 +431,26 @@ def add_serial(devices, name, filename):
...
+            :param filename: path to chardev filename
+            :param device_id: device id of the chardev
+            """
+            if not device_id:
+                device_id = "id_%s" % name
+            if chardev == "spicevmc":   # SPICE
+                cmd = " -chardev spicevmc"
+                cmd += ",id=%s,name=%s" % (device_id, name)
+            else:   # SOCKET
+                cmd = " -chardev socket"
```
Comment (suqinhuang) [inline virttest/qemu_vm.py]: need to support other backend as well, e.g file

**Comment by lmr** [comment]
Hi @xutian. As you might know, we're moving virt-test to avocado-vt, so I please ask you to consider re-implementing this functionality in the later. The new repo is:

https://github.com/avocado-framework/avocado-vt

Please keep in mind the following guidelines when creating the new PR:

1) Most of the time you can pick the patches of your original PR and merge them on a new branch with minimal changes.
3) Don't use autotest APIs, since we're moving away from using them. Please review the code and change things like `utils.system` -> `process.system`, `utils.run` -> `process.run`, so on and so forth.

Thanks, and I'm sorry for any inconveniences we might have caused your team.

**Pull request closed without merging**